# NeuroCamp Computer Science & Programming Day - Starter Notebook

1. Introduce the Jupyter Notebook, and run a "Hello World!" program in a notebook.
2. Variables and variable access.
3. I/O & load data from file.
4. Functions and modules. (We've used a function already!)
5. Loops and iteration.
6. Row field access. (Remember accessing values in a list by position or in a dictionary by key?)
7. Type conversions and template strings for output formatting.
8. Write formatted output to screen.
9. Math (i.e., change in measurement from A to B, percentage).
10. Conditionals.

## 1. Introduce the Jupyter Notebook, and run a "Hello World!" program in a notebook

This file that you are working with is called a Jupyter Notebook. It allows you to add and run code in locations within the file called "cells". Cells that run code will have square brackets in front of them, and may also have a cell number within them.

For our work, we'll be using the Python language (version 3.7, currently). If you look at the top right of the window, you should see "Python 3". This is the language indicator, and shows you what language the file is expected to use.

At the top left of this pane, you'll see the toolbar. That lets you control what's happening in the file and within the current cell. Most importantly, the right-most dropdown allows you to change a cell from "Markdown" explanatory text mode to "Code" Python programming mode.

In the cell below, we use Python 3's built-in `print` function to write `Hello, World!` to the output console (displayed below the cell when it is run). We'll talk more about functions later. For now, just make sure you understand the mechanics of how the notebook interface works.

**Feel free to play with the code cell below, and see if you can change the word `"World"` to display your name.**

In [11]:
print("Hello, World!")

Hello, World!


## 2. Variables and variable access

One thing we do frequently when writing programs is create variables to store data in memory that we can retrieve later. You can think of a "variable" as a box in which you put a piece of data called a "value".

You may have seen variables in algebra. When we do math, variables can only hold numbers.

When we write programs, the data that variables hold can have different types:

* text (called "strings")
* numbers (which can be "integer", "float", "double" and more)
* true/false (called "boolean")
* other types of single values
* structured and hierarchical data, such as lists, mappings (called "Dictionaries") and more (We'll see some of those later.)

Some languages require a different kind of box for each type of data you might want to put in. Python is gentler and allows us to put arbitrary types of data into the same variable. However, once you put a type of data into the variable, Python expects you to follow the rules of that type of data.

In the next cell, we're creating a couple string variables and then printing their values. Notice that the `print` function can accept variables representing the text ("string" value) to be written to the output.

**Once again, you should play with the code below and see if you can change the variable values.**

In [10]:
greeting = "Hello"
name = "World"

print(greeting + ", " + name + "!")

greeting = "Bienvenue"
name = "Arabella"

print(greeting + ", " + name + "!")

Hello, World!
Bienvenue, Arabella!


As you see, we can use variables in "expressions", like we did when we wrote `greeting + ", " + name + "!"`. That expression performs an operation called "concatenation" on the text given to it, providing us with a new string value that has all the text smashed together.

**_Aside_**: Note that we don't actually need the last `print` statement, since here in a notebook, entering only an expression (or even just a variable name by itself) as the last statement in a cell will evaluate that expression and print the value to the output area, as below. (Note, however, that the same is not true for values not at the end of the cell.)

In [8]:
greeting = "Hello"
name = "World"

greeting + ", " + name + "!"

greeting = "Bienvenue"
name = "Arabella"

greeting + ", " + name + "!"

'Bienvenue, Arabella!'

Expressions with number values also allow us to do mathematics, and more generally, expressions allow for manipulate the data stored in our variables.

In the below cell, we are creating two variables, then using them to create four more variables where the values are computed using expressions.

**Explore the code below to get a feel for how expressions work. For example, you might try to change the variable used in an expression, modify the mathematical operation used, or change the values of `a` and `b`.**

In [2]:
a = 1
b = 2
c = a + b
d = a - b
e = a * b
f = a / b

print("a =", a, ", b =", b, ", c =", c, ", d =", d, ", e =", e, ", f =", f)

a = 1 , b = 2 , c = 3 , d = -1 , e = 2 , f = 0.5


## 3. I/O & load data from file

When programming, "I/O" usually refers to "input/output", or the process of reading data from outside the program and delivering data outside the program.

We've already seen an I/O function: `print`. The `print` function is used to deliver string data from within the program to the screen, which lives outside the program.

Often, when working with I/O, we'll need to read from a file. In Python, files can be opened for reading using the `open` function, and lines can be read from the opened file using the `read` function.

Below, we read from our data file, located at `./data/data.csv` and print it to the screen.

In [4]:
opened_file = open('./data/data.csv')
contents = opened_file.read()
print(contents)
opened_file.close()

Subject,OSPAN_Group,OSPAN_Score,Distracter_easy,Distracter_hard
1,High,75,114.5,116.0
2,High,75,146.6,106.9
3,High,71,152.7,126.7
4,High,71,105.3,99.2
5,High,69,108.4,108.4
6,High,67,120.6,113.0
7,High,64,114.5,99.2
8,High,63,175.9,146.0
9,High,61,114.6,118.4
10,High,58,163.4,100.8
11,High,57,109.9,109.9
12,High,55,129.8,90.1
13,High,55,111.4,120.6
14,High,55,125.2,120.6
15,High,55,140.5,77.9
16,High,55,161.3,152.1
17,Low,52,120.6,123.7
18,Low,51,142.0,128.2
19,Low,51,117.6,129.8
20,Low,50,116.0,109.9
21,Low,48,151.1,152.7
22,Low,46,119.1,106.9
23,Low,41,142.0,135.9
24,Low,38,116.1,107.7
25,Low,38,135.9,109.9
26,Low,36,148.1,117.6
27,Low,22,163.4,154.2
28,Low,20,120.6,106.9
29,Low,19,106.1,107.7



This is great! We can make it a bit better, though.

Programmers often implement tools in computer languages to help them avoid problems. In this case, we have to remember to close the file after we're finished reading data from it. If we forget to do that, it can cause problems later on when trying to read the file, or even within the program.

To avoid this problem, the Python developers added a convenient way to write this same code that automatically closes the file when done. This uses something called "block syntax", which means that there is a collection of statements that are indented that represent all the things we want to do with the file while it is open. The block is started with a `with` statement.

In [5]:
with open('./data/data.csv') as opened_file:
    contents = opened_file.read()
    print(contents)

Subject,OSPAN_Group,OSPAN_Score,Distracter_easy,Distracter_hard
1,High,75,114.5,116.0
2,High,75,146.6,106.9
3,High,71,152.7,126.7
4,High,71,105.3,99.2
5,High,69,108.4,108.4
6,High,67,120.6,113.0
7,High,64,114.5,99.2
8,High,63,175.9,146.0
9,High,61,114.6,118.4
10,High,58,163.4,100.8
11,High,57,109.9,109.9
12,High,55,129.8,90.1
13,High,55,111.4,120.6
14,High,55,125.2,120.6
15,High,55,140.5,77.9
16,High,55,161.3,152.1
17,Low,52,120.6,123.7
18,Low,51,142.0,128.2
19,Low,51,117.6,129.8
20,Low,50,116.0,109.9
21,Low,48,151.1,152.7
22,Low,46,119.1,106.9
23,Low,41,142.0,135.9
24,Low,38,116.1,107.7
25,Low,38,135.9,109.9
26,Low,36,148.1,117.6
27,Low,22,163.4,154.2
28,Low,20,120.6,106.9
29,Low,19,106.1,107.7



It's great that we can see all the data, but right now, we're getting one variable `contents` that contains the entire file. Since this is tabular data, it would be nice to be able to get each row separately. For this, we need to look at a couple data structures: lists and Dictionaries, and to learn about...

## 4. Storing data

The list and dictionary data types were mentioned briefly above. Now we'll look at them more closely.

### Lists (also known as "Arrays")

A list in Python is an ordered collection of data. List values are written in Python as comma-separated lists enclosed in square brackets.

For example, in the next cell, we're defining a list containing the first 10 positive integers, and assigning that list-type value to the variable `my_list`.

In [7]:
my_list = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

my_list

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

Lists can contain other types of data. For example, the following is a list of string values.

In [10]:
['My', 'friend', 'ate', 'a', 'hot', 'dog', 'for', 'lunch']

['My', 'friend', 'ate', 'a', 'hot', 'dog', 'for', 'lunch']

If needed, we can put a variety of data types in the same list.

In [13]:
['This', 'list', 'has', 0, 'and', 8, 'and', False]

['This', 'list', 'has', 0, 'and', 8, 'and', False]

Values in a list are called "elements." To **access** the elements of a list, we can use the "index" of the element, which is a number representing the value's position within the list, starting from `0`.

For example, using `my_list` defined above, we can expect that the index `0`, our starting point, represents the first value in the list, `1`. Similarly, the index 5 refers to the fifth value in the list, which is `6`.

Let's see this in action.

**Play with the code below to explore lists. Here are a few things you might try:**

* What happens if you ask for the element at an index that is bigger than the length of the list?
* What happens if you ask for the element at an index that is negative?
* How would you retrieve the value 8 from the list?

In [22]:
my_list[4]

5

Were you able to answer the questions? OK! Let's apply this then.

**Now write a piece of code below that inspects the given array and reports the 3rd, 6th, and 9th elements by printing them to the screen.**

**Then, create a list of your top 3 favorite foods.**

In [32]:
words = [
    'This',
    'is',
    'my',
    'wonderful',
    'pet',
    'python',
    'Guido',
    'who',
    'enjoys',
    'life',
    'greatly'
]

In [34]:
# Enter your code here

# A. Print the 3rd, 6th and 9th elements of the given list of words to the output.

# B. Create a new list of your top 3 favorite foods, and print that to the output.


We can also modify list values after creation. To do this, we first select the element we want to change (as we did above), then use the equals sign to assign a new value to the list element.

In [36]:
words[3] = 'amazing'

words

['This',
 'is',
 'my',
 'amazing',
 'pet',
 'python',
 'Guido',
 'who',
 'enjoys',
 'life',
 'greatly']

There are many more ways to interact with lists. To learn more about lists and how they work, you can review [the Python tutorial section for them here](https://docs.python.org/3/tutorial/introduction.html#lists), and [the more detailed documentation here](https://docs.python.org/3/tutorial/datastructures.html#more-on-lists).

### Tuples (like lists, but "immutable")

Variable types that allow the value to be changed after creation, like lists and strings and integers, are called "mutable" types.

There are also usually similar types that do not allow the value to be changed after creation. These are called "immutable" types, and in general, are referred to as "constants" rather than "variables". We'll look at one of those now.

An immutable type related to the python list is a tuple. Tuples are created the same way we create lists, but using parentheses instead of square brackets.

In [37]:
('my', 'tuple', 'value')

('my', 'tuple', 'value')

In [38]:
a_tuple = (2, 'values', 'here', False)

print(a_tuple)
print(a_tuple[2])

# Trying to change the value will generate an error
a_tuple[0] = 4

(2, 'values', 'here', False)
here


TypeError: 'tuple' object does not support item assignment

Tuples are sometimes used when creating dictionaries, which we'll look at next.

As with lists, there's much more we can do with tuples. To learn more about tuples and how they work, you can review [the Python tutorial section for them here](https://docs.python.org/3/tutorial/datastructures.html#tuples-and-sequences).

### Dictionaries (also known as "Hash Maps")

Dictionaries are data structures that associate a collection of keys to corresponding values.

For example, in the next cell, we are creating a dictionary that associates the key `'key_one'` with the value `'some value'` and the key `'key_two'` with the value `'some other value'`.

In [35]:
my_dictionary = {
    'key_one': 'some value',
    'key_two': 'some other value'
}

my_dictionary

{'key_one': 'some value', 'key_two': 'some other value'}

As with lists, dictionaries can contain different types of data, both for values and for keys, however all keys must use immutable variable types.

In [46]:
{
    3: True,
    'key': 2.71828,
    (1, 'this', False): {
        'another': 'dictionary',
        'as a': 'value',
        'with': 'a tuple',
        'as': 'a key',
        'but': 'not',
        'a list': 'as a key'
    },
    'lists': ['can', 'be', 'values', True]
}

{3: True,
 'key': 2.71828,
 (1, 'this', False): {'another': 'dictionary',
  'as a': 'value',
  'with': 'a tuple',
  'as': 'a key',
  'but': 'not',
  'a list': 'as a key'},
 'lists': ['can', 'be', 'values', True]}

In practice, however, dictionaries are most often used to look up values or functions, and because of that, keys are usually just string values.

To retrieve a value from a dictionary, we can use the key we associated with that value. For example, to retrieve the value `'some value'` from `my_dictionary` defined above, we would use `'key_one'`.

In [47]:
my_dictionary['key_one']

'some value'

We can also change the value associated with a key in a dictionary in the same way we do with lists.

For example, to change the value associated with `'key_two'` in `my_dictionary`, we first select that value, then use the equals sign to reassign a new value to the same key.

In [49]:
my_dictionary['key_two'] = 'some new value'

my_dictionary

{'key_one': 'some value', 'key_two': 'some new value'}

Since the keys in a dictionary are not just counting numbers, we may need to see what they are. We can do that by using the dictionary's `keys` function. (As an aside, functions on objects like a dictionary are usually referred to as "methods".)

In [50]:
my_dictionary.keys()

dictionary_keys(['key_one', 'key_two'])

Although less useful, we can also get the values of a dictionary using its `values` method, and its key-value pairs using the `items` method.

In [52]:
my_dictionary.values()

dictionary_values(['some value', 'some new value'])

In [54]:
my_dictionary.items()

dictionary_items([('key_one', 'some value'), ('key_two', 'some new value')])

As with lists and tuples, there is much, much more we can do with dictionaries. To learn more about dictionaries and how they work, you can review [the Python tutorial section for them here](https://docs.python.org/3/tutorial/datastructures.html#dictionaries).

### Sorting data summary
A list, or an array, is an ordered collection of data. List values are written as comma-separated lists enclosed in square brackets.

`my_list = [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 ]`

A tuple is an immutable (unchangeable) collection of data enclosed in parentheses.

`my_tuple = (2, 'values', 'here', False)`

Dictionaries are data structures that associate a collection of keys to corresponding values. These are enclosed in curly brackets and are composed of keys and associated values.

`my_dictionary = {
    'key_one': 'some value',
    'key_two': 'some other value'
}`

**Play with the code below to explore dictionaries. Here are a few things you might try:**

* What happens if you ask for the element at using a key that is not in the dictionary?
* How would you retrieve the list of company divisions from the dictionary?
* How might you retrieve Bob Jones' telephone number from the dictionary?

In [72]:
company = {
    'divisions': [
        'Human Resources',
        'Research & Development',
        'Design',
        'Engineering',
        'Sales',
        'Marketing',
        'Finance',
        'Legal'
    ],
    'employees': [
        {
            'id': 125232,
            'name': 'Jenny Martin',
            'position': 'Sr. Software Engineer',
            'salary': 165000,
            'division': 'Engineering'
        },
        {
            'id': 653943,
            'name': 'Bob Jones',
            'position': 'Jr. Software Engineer',
            'salary': 59200,
            'division': 'Engineering'
        },
        {
            'id': 653943,
            'name': 'Althea Raines',
            'position': 'Chief Marketing Officer',
            'salary': 596400,
            'division': 'Sales'
        }
    ]
}

In [70]:
# Enter your code here


## 5. Functions and modules (we've used a function already!)

Now that we know about these data structures, let's look at how to package our code into reusable chunks.

When we wrote the statement `print(contents)` above to see the contents of the file, we were calling a function. A function is a reusable, packaged set of code statements, potentially with some configuration parameters, that can be executed and may return a value. Executing a function with some data is called "calling" the function with "arguments".

Let's say we have a collection of statements that takes some numbers and provides a tuple of their sum and difference:

In [55]:
a = 3
b = 4

a_plus_b = a + b
a_minus_b = a - b
(a_plus_b, a_minus_b)

(7, -1)

To turn this into a function, we turn that code into a block like we did with the `with` statement above. This time, we'll use `def` to define a function. After `def`, we need to give the function a name, which we'll use to call it later. Following that, we need to specify the arguments (and optionally their types and default values). The statement ends in a colon (`:`), which tells Python to look for a block of code after that line.

In the block, we need to put the `return` keyword before the tuple to tell Python that we want to pass this value back once the function's execution is done.

Finally, we can call the new function using its name, and adding the values for `a` and `b` in parentheses.

In [57]:
def plus_and_minus( a:int, b:int ):
    a_plus_b = a + b
    a_minus_b = a - b
    return (a_plus_b, a_minus_b)

plus_and_minus(3, 4)

(7, -1)

**Try writing a function on your own. Create a function that takes 3 inputs (named whatever you like), and computes the difference of the first and second inputs divided by the third. Be sure to return that computed value. Then, try calling that function with a few sets of numbers to verify it does what you expect.**

In [73]:
# Enter your code here

# A. Define the function

# B. Call the function with several sets of arguments


To learn more about functions and the variety of options you have when writing them, review [the tutorial section on them here](https://docs.python.org/3/tutorial/controlflow.html#defining-functions).

Often, the amount of code in a project gets quite large, and some of that code ends up being used in a number of places. To make things more manageable, developers often put that code into functions and call those where needed. If that still gets complex to manage, a developer can create a "module", sometimes also called a "package".

A module or package is a collection of Python code that is created in its own project, and which we can import into our project when needed. The Python language developers often also use packages to provide language features that may not always be needed.

Conveniently, the Python language developers provided a package for working with CSV data files. You can review [the documentation for that package here](https://docs.python.org/3/library/csv.html).

To import that package, we use the `import` keyword.

In [58]:
import csv

We can now rewrite the code above to use the CSV package. The package provides a `reader` function (called with `csv.reader`). Let's see what that gives us.

In [60]:
with open('./data/data.csv') as opened_file:
    contents = csv.reader(opened_file)
    print(contents)

That doesn't help much. Thankfully, Python also provides a package for inspecting objects in a program. Let's see what that tells us.

In [66]:
import inspect

with open('./data/data.csv') as opened_file:
    contents = csv.reader(opened_file)
    print(contents)
    print('-----------')
    print(inspect.getdoc(contents))

-----------
CSV reader

Reader objects are responsible for reading and parsing tabular data
in CSV format.


This provides a bit more information, but not much. At this point, we [go to the documentation](https://docs.python.org/3/library/csv.html), and we see the following:

```
>>> import csv
>>> with open('eggs.csv', newline='') as csvfile:
...     spamreader = csv.reader(csvfile, delimiter=' ', quotechar='|')
...     for row in spamreader:
...         print(', '.join(row))
Spam, Spam, Spam, Spam, Spam, Baked Beans
Spam, Lovely Spam, Wonderful Spam
```

This shows us that we have to use a `for` statement, which we haven't seen yet. That brings us to...

## 5. Loops and iteration

In computer science, the term "iteration" refers to making the computer repeat the same set of instructions multiple times. More commonly, a piece of code that implements an iteration is referred to as a "loop". While there are many kinds of loops, here, we'll look at the for loop.

In Python, a for loop begins with the keyword `for`, followed by an expression using the `in` keyword that defines the loop variable (`row` in the documentation example above), and the collection of data to select values from for the loop variable (`spamreader` in the documentation example above). The statement ends in a colon `:`, indicating to Python to expect a block following the statement. The block is the set of instructions to repeat for each value in the collection of data. In the documentation example above, the block contains a single statement: `print(', '.join(row))`.

In the code cell below, we've written a loop that iterates over the values in the tuple `(1, 4, 6, 5, 2)`, adds 3 to each, and prints the mathematical expression.

**Explore the code below to better understand how loops and iteration work. What happens when you...**

* Change the tuple to a list?
* Modify the statements in the block?

In [69]:
for num in (1, 4, 6, 5, 2):
    sum_value = num + 3
    print(num, "+ 3 =", sum_value)

1 + 3 = 4
4 + 3 = 7
6 + 3 = 9
5 + 3 = 8
2 + 3 = 5


There's much more we can learn about iteration, and Python has several unique features related to this concept, including generator expressions and list comprehensions, which are very powerful. To learn more about this, review [the Python tutorial section on for loops](https://docs.python.org/3/tutorial/controlflow.html#for-statements), [the more detailed Python documentation section on for statements](https://docs.python.org/3/reference/compound_stmts.html#for), and [the Python documentation on generators and generator expressions](https://docs.python.org/3/tutorial/classes.html#generators).

Let's rewrite our CSV file reading code to use the for loop.

In [74]:
with open('./data/data.csv') as opened_file:
    contents = csv.reader(opened_file)
    for row in contents:
        print(row)

['Subject', 'OSPAN_Group', 'OSPAN_Score', 'Distracter_easy', 'Distracter_hard']
['1', 'High', '75', '114.5', '116.0']
['2', 'High', '75', '146.6', '106.9']
['3', 'High', '71', '152.7', '126.7']
['4', 'High', '71', '105.3', '99.2']
['5', 'High', '69', '108.4', '108.4']
['6', 'High', '67', '120.6', '113.0']
['7', 'High', '64', '114.5', '99.2']
['8', 'High', '63', '175.9', '146.0']
['9', 'High', '61', '114.6', '118.4']
['10', 'High', '58', '163.4', '100.8']
['11', 'High', '57', '109.9', '109.9']
['12', 'High', '55', '129.8', '90.1']
['13', 'High', '55', '111.4', '120.6']
['14', 'High', '55', '125.2', '120.6']
['15', 'High', '55', '140.5', '77.9']
['16', 'High', '55', '161.3', '152.1']
['17', 'Low', '52', '120.6', '123.7']
['18', 'Low', '51', '142.0', '128.2']
['19', 'Low', '51', '117.6', '129.8']
['20', 'Low', '50', '116.0', '109.9']
['21', 'Low', '48', '151.1', '152.7']
['22', 'Low', '46', '119.1', '106.9']
['23', 'Low', '41', '142.0', '135.9']
['24', 'Low', '38', '116.1', '107.7']
['25'

This is much better! Now we see that the CSV package's `reader` function provides us with each row of the file as a list of string values.

## 6. Row field access (remember accessing values in a list by position or in a dictionary by key?)

Let's take each row apart, and print the values in the row with a vertical bar inbetween.

**Use what you've learned about lists to print each value in the `row` array separately, separated by the vertical bar character `|`.  For example, the row `['4', 'High', '71', '105.3', '99.2']` would print as:**

**`4 | High | 71 | 105.3 | 99.2`**

In [75]:
# Write code that prints each row's values separated by |.


This is nice, but referring to those values by their positions is a bit of a pain. It would be really nice to be able to refer to each row's column values by column name. This would require the CSV package to provide us a dictionary for each row. 

**Explore [the CSV package documentation](https://docs.python.org/3/library/csv.html) to see if you can find something that will do that.**

In [76]:
# Rewrite the code here using what you find in the CSV package documentation.


## 7. Type conversions and template strings for output formatting

One thing you may have noticed is that not all the vertical bars line up. There are sevreal ways to fix this problem, but the most common one is to use string formatting mechanisms built into the language to format the values appropriately.

Before we start, though, it's important to make sure we have the data in variables using the appropriate types. We can see from our prior loop code that each row list has every value as a string type. Since most of our data are numbers, this will be a problem if we want to compute mathematical expressions using the data (which we'll do shortly).

In order to fix this problem, we'll use Python's type conversion functions, which are just the names of the types.

**Let's start by exploring the types of various values. Explore this by using the `type` function with other values below.**

In [77]:
print('some text', ':', type('some text'))

print(345, ':', type(345))

print(2.718, ':', type(2.718))

print(True, ':', type(True))

some text : <class 'str'>
345 : <class 'int'>
2.718 : <class 'float'>
True : <class 'bool'>


To convert a string to another type, you can use the name of the type class (i.e., `int` for integer numeric values) as a function.

In [79]:
val = '125'
converted = int(val)
print('before:', type(val), ', after:', type(converted))

val = '3.7'
converted = float(val)
print('before:', type(val), ', after:', type(converted))

before: <class 'str'> , after: <class 'int'>
before: <class 'str'> , after: <class 'float'>


Note that if you try to convert to an incompatible type, for example, converting the text 3.7, which represents a floating point number, to an integer value, Python will generate an error.

In [81]:
val = '3.7'
converted = int(val)
print('before:', type(val), ', after:', type(converted))

ValueError: invalid literal for int() with base 10: '3.7'

Converting an actual floating point type to an integer is a valid operation, but be aware that Python will remove any fractional part of the number during conversion, meaning you will lose precision in your data.

In [82]:
val = 3.7
converted = int(val)
print('before:', val, type(val), ', after:', converted, type(converted))

before: 3.7 <class 'float'> , after: 3 <class 'int'>


**Explore type conversion in the cell below to be sure you understand it. Some things you may want to try:**

* Convert an integer value into a floating point value
* Convert a true/false value (type `bool`) into an integer value
* Convert a true/false value into a string value

In [84]:
# Explore type conversions here using the `int`, `float`, `str`, `bool`, and `type` functions


Let's rewrite our CSV reading code above to fix the types of the row values. Looking at the data, it seems we have columns with the following types:

* `Subject`: integer (`int`)
* `OSPAN_Group`: string (`str`)
* `OSPAN_Score`: integer (`int`)
* `Distracter_easy`: floating point (`float`)
* `Distracter_hard`: floating point (`float`)

Also, notice that the first row will not be able to be converted, as it has the column headings in it. We'll need to figure out how to handle that.

In [96]:
with open('./data/data.csv') as opened_file:
    contents = csv.DictReader(opened_file)
    cols = contents.fieldnames
    print(cols[0], '|', cols[1], '|', cols[2], '|', cols[3], '|', cols[4])
    for row in contents:
        print(int(row['Subject']), '|', row['OSPAN_Group'], '|', int(row['OSPAN_Score']),
              '|', float(row['Distracter_easy']), '|', float(row['Distracter_hard']))

Subject | OSPAN_Group | OSPAN_Score | Distracter_easy | Distracter_hard
1 | High | 75 | 114.5 | 116.0
2 | High | 75 | 146.6 | 106.9
3 | High | 71 | 152.7 | 126.7
4 | High | 71 | 105.3 | 99.2
5 | High | 69 | 108.4 | 108.4
6 | High | 67 | 120.6 | 113.0
7 | High | 64 | 114.5 | 99.2
8 | High | 63 | 175.9 | 146.0
9 | High | 61 | 114.6 | 118.4
10 | High | 58 | 163.4 | 100.8
11 | High | 57 | 109.9 | 109.9
12 | High | 55 | 129.8 | 90.1
13 | High | 55 | 111.4 | 120.6
14 | High | 55 | 125.2 | 120.6
15 | High | 55 | 140.5 | 77.9
16 | High | 55 | 161.3 | 152.1
17 | Low | 52 | 120.6 | 123.7
18 | Low | 51 | 142.0 | 128.2
19 | Low | 51 | 117.6 | 129.8
20 | Low | 50 | 116.0 | 109.9
21 | Low | 48 | 151.1 | 152.7
22 | Low | 46 | 119.1 | 106.9
23 | Low | 41 | 142.0 | 135.9
24 | Low | 38 | 116.1 | 107.7
25 | Low | 38 | 135.9 | 109.9
26 | Low | 36 | 148.1 | 117.6
27 | Low | 22 | 163.4 | 154.2
28 | Low | 20 | 120.6 | 106.9
29 | Low | 19 | 106.1 | 107.7


This is looking pretty good, but now we want to be able to format our output so that columns match up in a table.

To do this, we'll use Python's formatted "f-strings". To create an "f-string", we write text as if we're creating a regular string, but instead of enclosing the text in just quotes, we add an `f` in front of the starting quote mark. This tells Python that the string should be parsed for formatting. To add in our data, we can use curly braces `{` and `}` to create sections of the string that should be evaluated by Python for a value to be displayed.

An example of an "f-string" is below. Note that the data from `org` and the `expense` dictionary are used within the `{}` sections to specify what to print.

**Explore "f-strings" by printing various data with specific formatting**

In [116]:
org = "The church"
expense = {
    'purpose': 'building',
    'current_year_actual': 26952.94,
    'next_year_budgeted': 28500.00
}

message = f"{org} spent $ {expense['current_year_actual']} on {expense['purpose']} this year, and has budgeted $ {expense['next_year_budgeted']} for next year's expenses, a change of {(expense['next_year_budgeted'] - expense['current_year_actual']) / expense['current_year_actual']}."
print(message)

The church spent $ 26952.94 on building this year, and has budgeted $ 28500.0 for next year's expenses, a change of 0.05739856208636243.


This is good, and we're even doing a bit of math in the message, but we have some problems with the number of decimal places displayed, and it would be nice to have thousands separators on those dollar amounts.

To do this, we'll add formatting instructions to the `{}` sections to tell Python how to format that data.

Formatting instructions differ depending on the type of data and how you want to display it. For most work, we'll only be using a few formatting options. The most important are the field length and format indicator.

* For integer data, use the `d` format indicator.
* For floating point data, use the `f` indicator for fixed-point format, the `e` indicator for exponent notation, or the `g` indicator for smart switching between `f` and `e` depnding on the magnitude of the value.
* For percents represented as floating point values between 0 and 1, use the `%` indicator to automatically convert these to percentage values from 0% to 100%.

However, Python has a large number of formatting options available for "f-strings".  Rather than cover all that information here, take a moment to review [the format specification language here](https://docs.python.org/3/library/string.html#format-specification-mini-language) now.

Welcome back! Now that we have an idea of how to write formatting strings, let's look at the formatting examples below, and take them apart into their pieces. 

In [112]:
# floating point value
# 0,.2f ==> fixed format (f), variable length (0), ',' as thousands separator, 2 decimal places
cy_act = f"$ {expense['current_year_actual']:0,.2f}"
ny_bud = f"$ {expense['next_year_budgeted']:0,.2f}"

chg_val = (expense['next_year_budgeted'] - expense['current_year_actual']) / expense['current_year_actual']
# floating point value
# +0.2% ==> representing percentage, percent format (%), leading +/- sign (+), variable length (0), 2 decimal places
pct_chg = f"{chg_val:+0.2%}"

# Now all of these are already formatted strings, so we'll use the string formatting specifier (s) to format them.
message = f"{org:s} spent {cy_act:s} on {expense['purpose']:s} this year, and has budgeted {ny_bud:s} for next year's expenses, a change of {pct_chg:s}."
print(message)

The church spent $ 26,952.94 on building this year, and has budgeted $ 28,500.00 for next year's expenses, a change of +5.74%.


## 8. Write formatted output to the screen

Now that we know how to format data that we print to the screen, let's rework what we're printing out from the data file.

**Modify your previous CSV code using "f-strings" and format specifications to print a formatted table of values with aligned columns.**

In [113]:
with open('./data/data.csv') as opened_file:
    contents = csv.DictReader(opened_file)
    cols = contents.fieldnames
    # TODO: 3. Print the headers using a formatted f-string, ensuring column sizes are correct.
    print(cols[0], '|', cols[1], '|', cols[2], '|', cols[3], '|', cols[4])
    # TODO: 4. Add a horizontal rule of the correct length to separate headers from data.
    for row in contents:
        # TODO: 1. Create formatted strings, ensuring that fields are the correct size.
        # TODO: 2. Adjust below to print the new formatted strings.
        print(int(row['Subject']), '|', row['OSPAN_Group'], '|', int(row['OSPAN_Score']),
              '|', float(row['Distracter_easy']), '|', float(row['Distracter_hard']))


Subject | OSPAN_Group | OSPAN_Score | Distracter_easy | Distracter_hard
1 | High | 75 | 114.5 | 116.0
2 | High | 75 | 146.6 | 106.9
3 | High | 71 | 152.7 | 126.7
4 | High | 71 | 105.3 | 99.2
5 | High | 69 | 108.4 | 108.4
6 | High | 67 | 120.6 | 113.0
7 | High | 64 | 114.5 | 99.2
8 | High | 63 | 175.9 | 146.0
9 | High | 61 | 114.6 | 118.4
10 | High | 58 | 163.4 | 100.8
11 | High | 57 | 109.9 | 109.9
12 | High | 55 | 129.8 | 90.1
13 | High | 55 | 111.4 | 120.6
14 | High | 55 | 125.2 | 120.6
15 | High | 55 | 140.5 | 77.9
16 | High | 55 | 161.3 | 152.1
17 | Low | 52 | 120.6 | 123.7
18 | Low | 51 | 142.0 | 128.2
19 | Low | 51 | 117.6 | 129.8
20 | Low | 50 | 116.0 | 109.9
21 | Low | 48 | 151.1 | 152.7
22 | Low | 46 | 119.1 | 106.9
23 | Low | 41 | 142.0 | 135.9
24 | Low | 38 | 116.1 | 107.7
25 | Low | 38 | 135.9 | 109.9
26 | Low | 36 | 148.1 | 117.6
27 | Low | 22 | 163.4 | 154.2
28 | Low | 20 | 120.6 | 106.9
29 | Low | 19 | 106.1 | 107.7


Now, a bit of cleanup. We can print our table, but things are getting a little harder to read. Now is a good time to separate a piece of code that does a particular task into a function, then call it from here. Since all this formatting is what started the clutter in our code, let's create a new function that does the formatting of a row's data. Remember that we'll need to pass the row data as an argument to the function.

**Write a function that handles row formatting, and simplify the `with` statement's block by using your new function.**

In [117]:
# 1. TODO: Write a new function that handles formatting of a row. Your function can be named whatever you like,
#          but should take only one argument, a row of the data set (which is type `OrderedDict`).

# 2. TODO: Copy the code from your last effort above, and modify it to make use of your new function.


**Write another function that prints the header row and horizontal rule, then simplify your code once more by using this second new function.**

In [118]:
# 1. TODO: Write a new function that handles formatting and printing the table header. Your function can be named whatever you like,
#          but should take only one argument, the list of column names.

# 2. TODO: Copy the code from your immediately previous work, and modify it to make use of your new function.


## 9. Math (i.e., change in measurement from A to B, percentage)

Now that we've got our table printing nicely, we can look at doing some calculations on this data, and add the results as new columns to the table.

We've seen several times how to do mathematical calculations. These are similar to how you enter them in a calculator.

To add them as new columns in the result table, we'll need to do a few things for each new column:

* Create a function that does each calculation we're interested in.
* Within the iteration block, call that function to calculate the result, and add it to the row's dictionary object.
* Decide on header text for each calculation, and update our table header printing function.
* Update our row formatting function to add in the new calculated result.

**Add a new calculated column to your table showing the difference $Distracter_{easy} - Distracter_{hard}$.**

In [119]:
# 1. TODO: Write a new function that computes the desired difference value.

# 2. TODO: Copy all the code from above, including the row formatting function and the header printing
#          function. Modify them according to the last 3 bullets above.


**Add a new calculated column to your table showing the percent difference $\frac{Distracter_{easy} - Distracter_{hard}}{Distracter_{easy}}$.**

In [123]:
# 1. TODO: Write a new function that computes the desired percentage value. Hint: can you use the
#          difference computation function created above to help write the percent difference function?

# 2. TODO: Copy all the code from above, including the row formatting function, the header printing
#          function and the difference function. Modify them according to the last 3 bullets above.


To learn more about what Python offers for mathematical calculations, you can review [the tutorial section on using Python as a calculator](https://docs.python.org/3/tutorial/introduction.html#using-python-as-a-calculator) and [the tutorial introduction to Python's math package](https://docs.python.org/3/tutorial/stdlib.html#mathematics).

## 10. Conditionals

At times, we also want to be able to include values only if they belong to specific groups. For example, as we iterate through the rows of the data set, we may want to keep a count of the number of records in each `OSPAN_Group` and a total of all `OSPAN_Score`, `Distracter_easy` and `Distracter_hard` values for each group, so that we can calculate the average `OSPAN_Score` value for each `OSPAN_Group`.

When computers make decisions about what statements to execute, they use rules called "conditions", and such a decision is called a "conditional statement". The most common form of conditional statement looks like

```
if condition:
    block
else:
    block
```

This kind of statement allows for the developer to specify multiple possible blocks of statements depending on the value of some data.

The `condition` in an if-else statement must evaluate to either `True` or `False`. Conditions are usually written as logical assertions using comparison or boolean operators.

For example, if we want to check whether the `OSPAN_Group` value in a row is `'High'`, we might write a condition that looks like `row['OSPAN_Group'] == 'High'`. The `==` operator checks that the value on the left side is equal to the value on the right side. If we wanted to check that a value `my_number` is greater than 3, we could write `my_number > 3`.

If needed, if statements can omit the `else` section, or can have multiple condition checks using the `elif` keyword:

```
# No else
if condition:
    block

# Multiple conditions
if condition_1:
    block
elif condition_2:
    block
elif condition_3:
    block
...
else:
    block
```

You can read [more on if statements in the tutorial section here](https://docs.python.org/3/tutorial/controlflow.html#if-statements), and [more on conditions in the tutorial section on that here](https://docs.python.org/3/tutorial/datastructures.html#more-on-conditions)

**Use everything you've learned so far to add a footer to your table that has computed average values for each `OSPAN_Group` for the `Distracter_easy` and `Distracter_hard` columns.**

In [ ]:
# Good luck! Ask for help if you need it!


## Congratulations! You made it!

If you'd like to continue exploring programming for neuroscientific purposes, [open this notebook to work through implementing a simple 3-layer neural network](https://mybinder.org/v2/gh/loyno-mathcs/neurocamp_programming/master?filepath=NeuralNetwork.ipynb)!